_posts/2018-04-06-middleware.md

```diff
@@ -0,0 +1,207 @@
+    }
+}
+
+{% endhighlight %}
+
+As you can see, when dispatching the ExtractSimpleBusMessage event, you will need to pass the event object, and the DataContainer.
+
+{% highlight php%}
+// Dispatches the new ExtractSimpleBusMessage event
+try {
+    $this->eventDispatcher->dispatch(self::EXTRACT_SIMPLEBUS_MESSAGE, new ExtractSimpleBusMessage($message, $container));
+} catch (\Throwable $e) {
+    $this->log('alert', '', ['exception' => $e]);
+}
+{% endhighlight %}
+
+Below you can see the ExtractSimpleBusMessage class. The class in it self is very simple, it takes the message, and the
+DataContainer object as parameters, and adds some getters. This is what makes it possible for our decorators to access
+the DataContainer and add additional data to the message.
+
+{% highlight php %}
+<?php
+
+class ExtractSimpleBusMessage extends Event
+{
+    /** @var object */
+    private $simpleBusMessage;
+
+    /** @var DataContainer */
+    private $container;
+
+    public function __construct($simpleBusMessage, DataContainer $container)
+    {
+        $this->simpleBusMessage = $simpleBusMessage;
+        $this->container = $container;
+    }
+
+    public function getSimpleBusMessage()
+    {
+        return $this->simpleBusMessage;
+    }
+
+    public function getContainer(): DataContainer
+    {
+        return $this->container;
+    }
+}
+{% endhighlight %}
+
+## The DataContainer class
+Below is an example of how the DataContainer class could look like, as mentioned earlier, it's up to you to modify it
+to your needs.
+
+{% highlight php %}
+
+class DataContainer
+{
+    /** @var array */
+    private $data;
+
+    /**
+     * @param $data
+     */
+    public function __construct(array $data)
+    {
+        $this->data = $data;
+    }
+
+    public function addData(string $key, $value)
+    {
+        $this->data[$key] = $value;
+    }
+}
+
+{% endhighlight %}
+
+## Example of a decorator
+Now lets see an example of a decorator class.
+This example refers to when a new user has registered to your application.
+The decorator class listens to the ExtractSimpleBusMessage event. When it is dispatched, it will run the
+decorate() function.
+
+{% highlight php %}
+
+class UserHasRegisteredDecorator implements EventSubscriberInterface
+{
+    public static function getSubscribedEvents()
+    {
+        return [
+            Middleware::EXTRACT_SIMPLEBUS_MESSAGE => 'decorate',
+        ];
+    }
+    /**
+     * Populates the event with more specific data based on what is needed for the specific event.
+     *
+     * @param ExtractSimpleBusMessage $event
+     */
+    public function decorate(ExtractSimpleBusMessage $event)
+    {
+        // Retrieve the DataContainer object
+        $container = $event->getContainer();
+
+        // Check if the message and assigned event match
+        if ($event->getSimpleBusMessage() instanceof UserHasRegistered) {
+
+            // Add the data you want
+            $container->addData('age', '23');
```

Comment:
We have not explain why we would like to add this data. This example does not makes much sense...